src/Forms/GridField/GridState_Data.php

```diff
@@ -77,14 +96,56 @@ public function __toString()
         return json_encode($this->toArray());
     }
 
+    /**
+     * Convert the state to an array including only value that differ from the default state.
+     * @return array
+     */
     public function toArray()
     {
+        // If our current state is pristine, do not include any values
+        if (!$this->isDirty()) {
+            return [];
+        }
+
         $output = [];
 
         foreach ($this->data as $k => $v) {
-            $output[$k] = (is_object($v) && method_exists($v, 'toArray')) ? $v->toArray() : $v;
+            if (is_object($v) && method_exists($v, 'toArray')) {
+                $value = $v->toArray();
+                // Empty arrays represent pristine data, so we do not include them
+                if (empty($value)) {
+                    continue;
+                }
+            } else {
+                $value = $v;
+                // Check if we have a default value for this key and if it matches our current value
+                if (isset($this->defaults[$k]) && $this->defaults[$k] === $value) {
+                    continue;
+                }
+            }
+
+            $output[$k] = $value;
         }
 
         return $output;
     }
+
+    /**
+     * Whatever the current state differs from the stated default. If no default data has been provided, any data will
+     * be considered dirty.
+     * @return bool
+     */
+    public function isDirty(): bool
```

Comment:
What's the benefit of using `isDirty()` instead of relying on the content of toArray to do this work?